jicoco-websocket-client/src/test/kotlin/org/jitsi/websocket_client/WebSocketClientTest.kt

```diff
@@ -0,0 +1,120 @@
+ * limitations under the License.
+ */
+
+package org.jitsi.websocket_client
+
+import io.kotest.assertions.timing.eventually
+import io.kotest.core.spec.IsolationMode
+import io.kotest.core.spec.style.ShouldSpec
+import io.kotest.matchers.collections.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.types.shouldBeInstanceOf
+import io.ktor.client.HttpClient
+import io.ktor.client.engine.cio.CIO
+import io.ktor.client.features.websocket.WebSockets
+import io.ktor.http.cio.websocket.Frame
+import io.ktor.http.cio.websocket.readText
+import io.ktor.server.engine.embeddedServer
+import io.ktor.server.jetty.Jetty
+import org.jitsi.utils.logging2.LoggerImpl
+import kotlin.random.Random
+import kotlin.time.ExperimentalTime
+import kotlin.time.seconds
+
+@ExperimentalTime
+class WebSocketClientTest : ShouldSpec() {
+    override fun isolationMode(): IsolationMode? = IsolationMode.InstancePerLeaf
+
+    private val wsPort = Random.nextInt(1024, 65535).also {
+        println("Server running on port $it")
+    }
+
+    private val client = HttpClient(CIO) {
+        install(WebSockets)
+    }
+
+    private val wsServer = TestWsServer()
+
+    private val server = embeddedServer(Jetty, port = wsPort) {
+        wsServer.app(this)
+    }
+
+    private val receivedMessages = mutableListOf<Frame>()
+
+    private fun incomingMessageHandler(frame: Frame) {
+        receivedMessages.add(frame)
+    }
+
+    private val testLogger = LoggerImpl("test")
+
+    init {
+        server.start()
+
+        context("sendString") {
+            context("when no reply is expected") {
+                val ws = createAndStartWsClient("/ws/blackhole")
+                ws.sendString("hello")
+                should("send a message") {
+                    eventually(5.seconds) {
+                        wsServer.receivedMessages shouldHaveSize 1
+                    }
+                    wsServer.receivedMessages.first().shouldBeInstanceOf<Frame.Text>()
+                    (wsServer.receivedMessages.first() as Frame.Text).readText() shouldBe "hello"
+                }
+            }
+            context("when a reply is expected") {
+                val ws = createAndStartWsClient("/ws/echo")
+                ws.sendString("hello")
+                should("invoke the incoming message handler") {
+                    eventually(5.seconds) {
+                        receivedMessages shouldHaveSize 1
+                    }
+                    receivedMessages.first().shouldBeInstanceOf<Frame.Text>()
+                    (receivedMessages.first() as Frame.Text).readText() shouldBe "hello"
+                }
+            }
+            context("when the client is stopped while waiting for a reply") {
+                val ws = createAndStartWsClient("/ws/delayandclose")
+                Thread.sleep(1000)
+                ws.sendString("hello")
+                ws.stop()
+                Thread.sleep(1000)
```

Comment:
What does this test for, just lack of exceptions? Are both `sleep` calls necessary?